evalml/pipelines/components/transformers/samplers/undersampler.py

```diff
@@ -59,20 +78,98 @@ def _initialize_sampler(self, X, y):
+        counts = y.value_counts()
+        normalized_counts = y.value_counts(normalize=True)
+        minority_class_count = min(normalized_counts)
+        class_ratios = minority_class_count / normalized_counts
+        # if no class ratios are larger than what we consider balanced, then the target is balanced
+        if all(class_ratios >= self.sampling_ratio):
+            return {}
+        # if any classes have less than min_samples counts and are less than min_percentage of the total data,
+        # then it's severely imbalanced
+        if any(counts < self.min_samples) and any(
+            normalized_counts < self.min_percentage
+        ):
+            return {}
+        # otherwise, we are imbalanced enough to perform on this
+        undersample_classes = counts[class_ratios <= self.sampling_ratio].index.values
+        # find goal size, round it down if it's a float
+        minority_class = min(counts.values)
+        goal_value = max(
+            int((minority_class / self.sampling_ratio) // 1), self.min_samples
+        )
+        # we don't want to drop less than 0 rows
+        drop_values = {k: max(0, counts[k] - goal_value) for k in undersample_classes}
+        return {k: v for k, v in drop_values.items() if v > 0}
+
+    def _sampling_dict_to_remove_dict(self, y):
+        """Turns the sampling dict input into a dict of samples to remove for each target, similar to the return of _find_ideal_samples.
+
+        Arguments:
+            y (pd.Series): Training data targets
+
+        Returns:
+            (dict): dictionary with undersample target class as key, and number of samples to remove as the value.
+                If we don't need to resample, returns empty dictionary.
+        """
+        y_dict = y.value_counts().to_dict()
+        new_dic = {}
+        for k, v in self.sampling_ratio_dict.items():
+            new_dic[k] = max(y_dict[k] - v, 0)
+        return new_dic
+
+    def fit_resample(self, X, y):
+        """Resampling technique for this sampler.
+
+        Arguments:
+            X (pd.DataFrame): Training data to fit and resample
+            y (pd.Series): Training data targets to fit and resample
+
+        Returns:
+            list: Indices to keep for training data
+        """
+
+        if self.parameters["sampling_ratio_dict"]:
+            self.sampling_ratio_dict = self._convert_dictionary(
+                self.parameters["sampling_ratio_dict"], y
+            )
+
+        y = infer_feature_types(y)
+
+        if len(self.sampling_ratio_dict):
+            result = self._sampling_dict_to_remove_dict(y)
+        else:
+            result = self._find_ideal_samples(y)
+        indices_to_drop = []
+        if len(result):
+            # iterate through the classes we need to undersample and remove the number of samples we need to remove
+            for key, value in result.items():
+                indices = y.index[y == key].values
+                # import pdb; pdb.set_trace()
```

Comment:
Can we get rid of this while we're here?  I would blame this on me.

Reply:
LOL oops